### PR TITLE
CNV-5076 Exposing next run configuration to be applied after VM restart

### PR DESCRIPTION
**Pull request by sjhala-ccs**
This PR addresses https://issues.redhat.com/browse/CNV-5076 

If the virtual machine is running, changes to select fields in the UI will be displayed as pending changes and take effect after the virtual machine is restarted. Added a note about this feature in the following existing modules:

- **Details tab** (Preview build: https://cnv-5076--ocpdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/virt-edit-vms.html#virt-editing-vm-web_virt-edit-vms)

- **Boot Order** (Preview build: https://cnv-5076--ocpdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/virt-edit-boot-order.html)

- **Network Interfaces tab** (Preview build: https://cnv-5076--ocpdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/virt-edit-vms.html#virt-vm-add-nic_virt-edit-vms)

- **Disks tab** (Preview build: https://cnv-5076--ocpdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/virt-edit-vms.html#virt-vm-add-disk_virt-edit-vms)

- **Environment tab** (Preview build: https://cnv-5076--ocpdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/virt-managing-configmaps-secrets-service-accounts.html#virt-adding-secret-configmap-service-account-to-vm_virt-managing-configmaps-secrets-service-accounts)

Code review requested from Tomas Jelinek and Ido Rosenzwig (tagged in Jira)
QE review requested from Nelly Credi and Guohua Ouyang (tagged in Jira)

Peer review needed
Merge/CP to branch/enterprise-4.6

### CODE REVIEW

File: modules/virt-adding-secret-configmap-service-account-to-vm.adoc [hunk 1]
```diff
@@ -16,11 +16,11 @@ Add a secret, ConfigMap, or service account to a virtual machine by using the {p
```
Comment (sjhala-ccs) [inline modules/virt-adding-secret-configmap-service-account-to-vm.adoc:16]: The existing documentation states that the VM must be powered off as a prerequisite to making any changes in the Environment tab. Is that still accurate? If yes, then since the VM is not running there would not be any question of pending changes to be applied after restart.

File: modules/virt-removing-secret-configmap-service-account-vm.adoc [hunk 2]
```diff
@@ -17,11 +17,11 @@ that is attached to a virtual machine.
```
Comment (sjhala-ccs) [inline modules/virt-removing-secret-configmap-service-account-vm.adoc:17]: See my question above re: Environment tab.

File: modules/virt-remove-boot-order-item-web.adoc [hunk 3]
```diff
@@ -21,4 +21,12 @@ Remove items from a boot order list by using the web console.
 
 . Click the pencil icon that is located on the right side of *Boot Order*.
 
-. Click the *Remove* icon next to the item. The item is removed from the boot order list and saved in the list of available boot sources. If you remove all items from the boot order list, the following message displays: *No resource selected. VM will attempt to boot disks from YAML by order of appearance in YAML file. Please select a boot source.*
+. Click the *Remove* icon next to the item. The item is removed from the boot order list and saved in the list of available boot sources. If you remove all items from the boot order list, the following message displays: *No resource selected. VM will attempt to boot disks from YAML by order of appearance in YAML file.*
```
Comment (irosenzw) [inline modules/virt-remove-boot-order-item-web.adoc:24]: maybe rephrase the last sentence:
`The VM will attempt to boot the disks from YAML by order of appearance.`

**Review by irosenzw — COMMENTED** [review]
First all of nice job with documenting this feature. I've added small things which I find important to add.
In addition to what I wrote, I think it's worth mentioning the info and warning alerts in the dialogs (boot order, flavor, network interfaces and disks) while the VM is running. Explaining the user he is being notified when he is about to make a change to a property that will require him to restart VM in order to apply the change.

In general, it would be nice if you can add screenshots of these messages and alerts we added in this feature. For example you mention many times a pencil icon, it would be nice showing it.
Moreover, add an example of the `Pending Changes` alert at top of the screen. etc...

File: modules/virt-add-boot-order-web.adoc [hunk 4]
```diff
@@ -18,10 +18,18 @@ Add items to a boot order list by using the web console.
 
 . Click the *Details* tab.
 
-. Click the pencil icon that is located on the right side of *Boot Order*. If a YAML configuration does not exist, or if this is the first time that you are creating a boot order list, the following message displays: *No resource selected. VM will attempt to boot disks from YAML by order of appearance in YAMLv file. Please select a boot source*.
+. Click the pencil icon that is located on the right side of *Boot Order*. If a YAML configuration does not exist, or if this is the first time that you are creating a boot order list, the following message displays: *No resource selected. VM will attempt to boot disks from YAML by order of appearance in YAMLv file.*
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-add-boot-order-web.adoc:21]: Fixed the typo.

File: modules/virt-remove-boot-order-item-web.adoc (hunk 3 above)
Comment (sjhala-ccs) [reply modules/virt-remove-boot-order-item-web.adoc:24]: Done

File: modules/virt-adding-secret-configmap-service-account-to-vm.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-adding-secret-configmap-service-account-to-vm.adoc:16]: Okay. Will remove.

File: modules/virt-add-disk-to-vm.adoc [hunk 5]
```diff
@@ -24,22 +26,34 @@ Use this procedure to add a virtual disk to a {object}.
...
+. Click *Add*.
+
+ifdef::virt-vm[]
+[NOTE]
+====
+If the {object} is running, the new disk is in the *pending restart* state and changes will not take effect until you restart the {object}.
+
+A list of all pending changes, that will be applied when the {object} restarts, is displayed in the banner above the *Disks* tab.
+====
+endif::virt-vm[]
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-add-disk-to-vm.adoc:46]: The note mentions that the new disk will remain in the `pending restart` state until the virtual machine is restarted.

File: modules/virt-add-nic-to-vm.adoc [hunk 6]
```diff
@@ -21,36 +23,34 @@ Use this procedure to add a network interface to a {object}.
...
 
-The new network interface has a `Pending VM restart` Link State until you
-restart the virtual machine. Hover over the Link State to display more
-detailed information.
-
-The *Link State* is set to *Up* by default when the network interface card
-is defined on the virtual machine and connected to the network.
+A list of all pending changes, that will be applied when the {object} restarts, is displayed in the banner above the *Network Interfaces* tab.
+====
+endif::virt-vm[]
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-add-nic-to-vm.adoc:43]: The note mentions that the NIC will be in `pending restart` state until the virtual machine is restarted.

File: modules/virt-adding-secret-configmap-service-account-to-vm.adoc [hunk 7]
```diff
@@ -16,11 +16,11 @@ Add a secret, ConfigMap, or service account to a virtual machine by using the {p
...
-. From the side menu, click *Virtualization*.
+. Click *Workloads* -> *Virtualization* from the side menu.
 
-. Click the *Virtual Machine* tab.
+. Click the *Virtual Machines* tab.
 
-. Select a virtual machine to open its *Virtual Machine Overview* page.
+. Select a virtual machine to open the *Virtual Machine Overview* screen.
 
 . Click the *Environment* tab.
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-adding-secret-configmap-service-account-to-vm.adoc:24]: Incorporated your comments

**Comment by sjhala-ccs** [comment]
> In general, it would be nice if you can add screenshots of these messages and alerts we added in this feature. For example you mention many times a pencil icon, it would be nice showing it.
> Moreover, add an example of the `Pending Changes` alert at top of the screen. etc...

As a CCS team best practice, we typically do not add screenshots because UI elements can change frequently and the images are difficult to maintain.

File: modules/virt-add-disk-to-vm.adoc [hunk 8]
```diff
@@ -24,22 +26,34 @@ Use this procedure to add a virtual disk to a {object}.
 
 .Procedure
 
-. From the *{object-gui}* tab, select your {object}.
-. Select the *Disks* tab.
-. Click *Add Disks* to open the *Add Disk* window.
-. In the *Add Disk* window, specify *Source*, *Name*, *Size*, *Interface*, and *Storage Class*.
+. Click *Workloads* -> *Virtualization* from the side menu.
+. Click the *{object-gui}s* tab.
+. Select a {object} to open the *{object-gui} Overview* screen.
```
Comment (ousleyp) [inline modules/virt-add-disk-to-vm.adoc:31]: We say "screen" a lot in our docs, but the ISG says to only use "screen" if you're referring to the physical surface of a display. I think the instances of "screen" in this PR could all be changed to "page" but please correct me if I am wrong about that.

File: modules/virt-add-disk-to-vm.adoc [hunk 9]
```diff
@@ -24,22 +26,34 @@ Use this procedure to add a virtual disk to a {object}.
...
+. Click the *Disks* tab.
+. Click *Add Disk* to open the *Add Disk* window.
+. In the *Add Disk* window, specify *Source*, *Name*, *Size*, *Interface*, *Type*, and *Storage Class*.
```
Comment (ousleyp) [inline modules/virt-add-disk-to-vm.adoc:34]: s/specify/specify the

File: modules/virt-add-disk-to-vm.adoc [hunk 10]
```diff
@@ -24,22 +26,34 @@ Use this procedure to add a virtual disk to a {object}.
...
-. Use the drop-down lists and check boxes to edit the disk configuration.
-. Click *OK*.
+. Click *Add*.
+
+ifdef::virt-vm[]
+[NOTE]
+====
+If the {object} is running, the new disk is in the *pending restart* state and will not be attached until you restart the {object}.
+
+A list of all pending changes, that will be applied when the {object} restarts, is displayed in the *Pending Changes* banner on top of the page.
```
Comment (ousleyp) [inline modules/virt-add-disk-to-vm.adoc:43]: This sentence needs a couple of tweaks (passive voice, commas, + wording) so I have a rewrite suggestion for you to take or leave: 

`The *Pending Changes* banner at the top of the page displays a list of all changes that will be applied when the {object} restarts.`

(Applies to all instances of this sentence in the PR)

File: modules/virt-add-nic-to-vm.adoc [hunk 11]
```diff
@@ -21,36 +23,34 @@ Use this procedure to add a network interface to a {object}.
...
-. Click *Save Changes*.
-. Click *OK*.
+. Click *Add*.
 
-The new network interface displays at the top of the *Create Network Interface* list
-until the user restarts it.
+ifdef::virt-vm[]
+[NOTE]
+====
+If the {object} is running, the new network interface is in *pending restart* state and changes will not take effect until you restart the {object}.
```
Comment (ousleyp) [inline modules/virt-add-nic-to-vm.adoc:38]: s/is in/is in the

File: modules/virt-adding-secret-configmap-service-account-to-vm.adoc [hunk 12]
```diff
@@ -12,35 +12,42 @@ Add a secret, ConfigMap, or service account to a virtual machine by using the {p
...
 [NOTE]
 ====
-You can reset the form to the last saved state by clicking *Reload*.
-====
+.. You can reset the form to the last saved state by clicking *Reload*.
 
+.. The *Environment* resources are added to the virtual machine as disks. You can mount the secret, ConfigMap, or service account as you would mount any other disk.
+
+.. If the virtual machine is running, changes will not take effect until you restart the virtual machine. The newly added resources are marked as pending changes
+for both the *Environment* and *Disks* tab in the *Pending Changes* banner on top of the page.
```
Comment (ousleyp) [inline modules/virt-adding-secret-configmap-service-account-to-vm.adoc:39]: s/on top of/at the top of

File: modules/virt-adding-secret-configmap-service-account-to-vm.adoc [hunk 13]
```diff
@@ -12,35 +12,42 @@ Add a secret, ConfigMap, or service account to a virtual machine by using the {p
...
 .Verification steps
 
 . From the *Virtual Machine Overview* page, click the *Disks* tab.
 
 . Check to ensure that the secret, ConfigMap, or service account is included in the list of disks.
 
-. Optional. Start the virtual machine by clicking *Actions* -> *Start Virtual Machine*.
+. Optional. If the virtual machine is running, restart the virtual machine by clicking *Actions* -> *Restart Virtual Machine*.
+
+. Optional. If the virtual machine is stopped, start the virtual machine by clicking *Actions* -> *Start Virtual Machine*.
```
Comment (ousleyp) [inline modules/virt-adding-secret-configmap-service-account-to-vm.adoc:51]: Because the user won't need to both start and restart the VM, I think these should be condensed into one step with two bullet points underneath it for the two different scenarios.

File: modules/virt-edit-boot-order-web.adoc [hunk 14]
```diff
@@ -26,3 +26,11 @@ Edit the boot order list in the web console.
 * If you use a screen reader, press the Up Arrow key or Down Arrow key to move the item in the boot order list. Then, press the *Tab* key to drop the item in a location of your choice.
 
 . Click *Save*.
+
+[NOTE]
+====
+If the virtual machine is running, changes to the boot order list will not take effect until you restart the virtual machine.
+
+You can view pending changes by clicking *View Pending Changes* next to the pencil icon of the *Boot Order* field. A list of all pending changes, that will
```
Comment (ousleyp) [inline modules/virt-edit-boot-order-web.adoc:34]: Instead of "next to," can you specify which side of the pencil icon it's on (right or left)? This also applies to the virt-editing-vm-web module and the virt-remove-boot-order-item-web module.

**Review by ousleyp — COMMENTED** [review]
Overall, this is looking great. I have a few minor suggestions/nits which repeat throughout the PR. Let me know if you have any questions. Thanks, Shikha!

File: modules/virt-add-disk-to-vm.adoc (hunk 10 above)
Comment (sjhala-ccs) [reply modules/virt-add-disk-to-vm.adoc:43]: Changes applied to all instances of the sentence

File: modules/virt-add-disk-to-vm.adoc (hunk 8 above)
Comment (sjhala-ccs) [reply modules/virt-add-disk-to-vm.adoc:31]: Yes, we do say "screen" a lot in our docs in spite of what the ISG says. Those three steps appear in all web console procedures, not just in this PR. I think changing "screen" to "page" would involve a larger effort and a separate PR.

File: modules/virt-edit-boot-order-web.adoc (hunk 14 above)
Comment (sjhala-ccs) [reply modules/virt-edit-boot-order-web.adoc:34]: Done
